situp-plugins/elasticsearch/src/main/java/com/amazon/situp/plugins/sink/elasticsearch/IndexConfiguration.java

```diff
@@ -1,170 +1,218 @@
-  public static final String BULK_SIZE = "bulk_size";
-  public static final String DOCUMENT_ID_FIELD = "document_id_field";
-  public static final long DEFAULT_BULK_SIZE = 5L;
-
-  private final String indexType;
-  private final String indexAlias;
-  private final URL templateURL;
-  private final String documentIdField;
-  private final long bulkSize;
-
-  public String getIndexType() {
-    return indexType;
-  }
-
-  public String getIndexAlias() {
-    return indexAlias;
-  }
-
-  public URL getTemplateURL() {
-    return templateURL;
-  }
-
-  public String getDocumentIdField() {
-    return documentIdField;
-  }
-
-  public long getBulkSize() {
-    return bulkSize;
-  }
-
-  public static class Builder {
-    private boolean isRaw = false;
-    private boolean isServiceMap = false;
-    private String indexAlias;
-    private String templateFile;
-    private String documentIdField;
-    private long bulkSize = DEFAULT_BULK_SIZE;
-
-    public Builder setIsRaw(final Boolean isRaw) {
-      checkNotNull(isRaw, "trace_analytics_raw cannot be null.");
-      this.isRaw = isRaw;
-      return this;
-    }
+    /**
+     * TODO: add index management policy parameters
+     */
+    public static final String SETTINGS = "settings";
+    public static final String TRACE_ANALYTICS_RAW_FLAG = "trace_analytics_raw";
+    public static final String TRACE_ANALYTICS_SERVICE_MAP_FLAG = "trace_analytics_service_map";
+    public static final String INDEX_ALIAS = "index";
+    public static final String TEMPLATE_FILE = "template_file";
+    public static final String NUM_SHARDS = "number_of_shards";
+    public static final String NUM_REPLICAS = "number_of_replicas";
+    public static final String BULK_SIZE = "bulk_size";
+    public static final String DOCUMENT_ID_FIELD = "document_id_field";
+    public static final long DEFAULT_BULK_SIZE = 5L;
+
+    private final String indexType;
+    private final String indexAlias;
+    private final Map<String, Object> indexTemplate;
+    private final String documentIdField;
+    private final long bulkSize;
+
+    @SuppressWarnings("unchecked")
+    private IndexConfiguration(final Builder builder) {
+        final String indexType;
+        if (builder.isRaw && builder.isServiceMap) {
+            throw new IllegalStateException("trace_analytics_raw and trace_analytics_service_map cannot be both true.");
+        } else if (builder.isRaw) {
+            indexType = IndexConstants.RAW;
+        } else if (builder.isServiceMap) {
+            indexType = IndexConstants.SERVICE_MAP;
+        } else {
+            indexType = IndexConstants.CUSTOM;
+        }
+        this.indexType = indexType;
+
+        this.indexTemplate = readIndexTemplate(builder.templateFile, indexType);
+
+        if (builder.numReplicas > 0) {
+            indexTemplate.putIfAbsent(SETTINGS, new HashMap<>());
+            ((Map<String, Object>) indexTemplate.get(SETTINGS)).putIfAbsent(NUM_REPLICAS, builder.numReplicas);
```

Comment:
*Jus for my informaiton*
if there is already an existing template, we won't override ?
if there is already an existing replicas - we will not honor the builder value ?

Reply:
Correct. If users provide shards and replica as a part of the template file, we will not override.